docs: Added initial PR template with directions for doc only changes and squash merges [no ci]

nicolasperez19 opened the pull request:
# Summary
- Solves issue #7657

# Changes Made
- Added directions to the `README` so that collaborators follow the PR template when sending a PR
- Added a general pull request template
  - Suggests to collaborators to add `[no ci]` to the commit title
  - Suggests to collaborators to format commit titles `<module>:<commit title> (#<issue_number>)` when squashing multiple commits on merge

# Additional Information
- The general PR template is very rough, so adding changes is welcome!

nicolasperez19 committed:
doc: Added to README that collaborators should follow PR template

nicolasperez19 committed:
doc: Added general pull request template with directions for doc only changes and squash merges

mofosyne commented:
Nice, I be interested to see if everyone else think it's good.

I would like to also add that I've been adding 'review complexity' manually, but it be good to get the submitter to indicate their subjective take on how hard it would be to review their PR. The idea is that at the very least it would speed up my manual assessment and labeling of PR... and also that later we can add some automation to do the PR complexity label automagically.

Did a bit of research and found this option out:

* https://github.com/Bhacaz/label-regex
    - It looks to do the job but is not an official github action

We are currently using this for labeling PRs but it can only do it based on content and not on PR title and description.

* issue labeler by actions: https://github.com/actions/labeler
    - We are using this but its designed specifically for PR labeling based on file content and branch name, but does not have the ability to label by content.
        - Chasing up this PR https://github.com/actions/labeler/pull/688 which does enable such feature

In this case however to keep this PR complexity down, don't try to add this automation in. We just need to make sure we can add a regex matching process for it in the future somehow.

Approving, but waiting for others to see if we need to flesh it out. I think it's good to keep this simple however.

ggerganov commented on .github/PULL_REQUEST_TEMPLATE/pull_request_template.md:
```diff
@@ -0,0 +1,40 @@
+* **New Features/Changes**: [Describe new features or significant changes added in this PR]
+* **Fixed Issues**: [List specific issues or bugs fixed by these changes]
+
+## Verification
+
+To verify this PR, you can:
+
+* Run automated tests or scripts to ensure the changes do not introduce errors
+* Review code for style, security, and best practices
+* Verify that all changed files are properly formatted and consistent
+
```
Replace with actual test commands, such as `./tests/test-backend-ops` and reference to [ggml-ci](https://github.com/ggerganov/llama.cpp/tree/master/ci)

ggerganov commented on .github/PULL_REQUEST_TEMPLATE/pull_request_template.md:
```diff
@@ -0,0 +1,40 @@
+* Review code for style, security, and best practices
+* Verify that all changed files are properly formatted and consistent
+
+## Additional Information (Optional)
+
+[Add any additional context, explanations, or requests that are relevant to this PR.]
+
+**Important Notes**
+
+* If this pull request only contains documentation changes (e.g., updating
+READMEs, adding new wiki pages), please add `[no ci]` to the commit title.
+This will skip unnecessary CI checks and help reduce build times.
+* When squashing multiple commits on merge, use the following format for
+your commit title: `<module>:<commit title> (#<issue_number>)`. For example: `utils: Fix typo in utils.py (#1234)`
+* Please ensure that this PR follows our contributing guidelines, available at [](README.md). This includes formatting code according to our style guide and ensuring that all changes are thoroughly tested.
```
Should be moved to the CONTRIBUTING file

ggerganov reviewed (commented):
It's better to add a CONTRIBUTING file in the repo and the PR template to just reference the file:

```
Before opening the PR make sure to read CONTRIBUTING
```

nicolasperez19 committed:
doc: Moved contributing info from PR template to separate CONTRIBUTING file

nicolasperez19 committed:
docs: Added link in README to pull request template

nicolasperez19 committed:
doc: Simplified PR template, referenced ggml-ci and test commands

nicolasperez19 commented:
I made some modifications following everyone's advice. 

- [x] I added a field to the PR template for the review complexity level so that it's easier for @mofosyne to triage PRs. 
- [x] I moved most of the contribution guidelines to its own markdown file
- [x] In the PR template I referenced `ggml-ci` and the test commands in the `tests` folder, and gave an example command following @ggerganov's adivce

ggerganov commented on .github/PULL_REQUEST_TEMPLATE/pull_request_template.md:
```diff
@@ -0,0 +1,16 @@
+    * For instance, running the `./tests/test-backend-ops` command tests different backend implementations of the GGML library
+    * 
+
+* Execute [the full CI locally on your machine](ci/README.md) before publishing
```
I would like the PR template to be very short, basically just this:

```
[x] I have read CONTRIBUTING.md
```

The reason is that seasoned contributors already know how to make PRs and what to include in the description, so no need for much text here.

We can then move the Contributing and Coding Guidelines from the README to the CONTRIBUTING file

nicolasperez19 committed:
doc: Moved all PR template information to CONTRIBUTING.md

nicolasperez19 committed:
doc: Made PR template only contain a line that a collaborator has read the contributing guidelines

nicolasperez19 commented:
I took out most of the fluff from the PR template and moved it to CONTRIBUTING

compilade commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,14 @@
+* Please rate the complexity of your PR (i.e. `easy`, `medium`, `hard`). This makes it easier for maintainers to triage the PRs.
+* If the pull request only contains documentation changes (e.g., updating
+READMEs, adding new wiki pages), please add `[no ci]` to the commit title. This will skip unnecessary CI checks and help reduce build times.
+* When squashing multiple commits on merge, use the following format for your commit title: `<module>:<commit title> (#<issue_number>)`. For example: `utils: Fix typo in utils.py (#1234)`
```
I think suggesting to use spaces seems preferable, because that's how both [ggerganov](https://github.com/ggerganov/llama.cpp/pulls?q=is%3Apr+is%3Aclosed+author%3Aggerganov) and [slaren](https://github.com/ggerganov/llama.cpp/pulls?q=is%3Apr+is%3Aclosed+author%3Aslaren) format their PR titles and commit messages.
```suggestion
* When squashing multiple commits on merge, use the following format for your commit title: `<module> : <commit title> (#<issue_number>)`. For example: `utils : Fix typo in utils.py (#1234)`
```

mofosyne committed:
Update CONTRIBUTING.md [no ci]

Co-authored-by: compilade <git@compilade.net>

compilade commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,14 @@
+
+## PR formatting
+
+* Please rate the complexity of your PR (i.e. `easy`, `medium`, `hard`). This makes it easier for maintainers to triage the PRs.
```
This should probably reflect the label names

```suggestion
* Please rate the complexity of your PR (i.e. `low`, `medium`, `high`). This makes it easier for maintainers to triage the PRs.
```

mofosyne replied:
Agreed. I do wonder if it's reasonable to ask people to type something like `medium complexity` as that might be less ambiguous to match against a regex in the future.

mofosyne replied:
directly edited it just now so all good

ggerganov commented on README.md:
```diff
@@ -1088,6 +1088,7 @@ docker run --gpus all -v /path/to/models:/models local/llama.cpp:server-cuda -m
 
 - Contributors can open PRs
 - Collaborators can push to branches in the `llama.cpp` repo and merge PRs into the `master` branch
+- Collaborators should follow the [PR template](.github/PULL_REQUEST_TEMPLATE/pull_request_template.md) when adding a PR
```
This is not necessary anymore

mofosyne replied:
fixed

mofosyne committed:
Update README.md

slaren commented:
It would be nice if there was a brief explanation of `llama-bench`, `compare-llama-bench.py` and `compare-commits.sh` for PRs that affect performance. It's not very easy to find information about these tools without looking at other PRs, and many people are still using `main` to test performance. Basically, if the PR is meant to improve performance, it is recommended to include the results performance tests using `llama-bench` or its related tools.

mofosyne committed:
Update pull_request_template.md [no ci]

mofosyne committed:
Update CONTRIBUTING.md to clarify review complexity labeling [no ci]

mofosyne commented:
@slaren I'm not sure @nicolasperez19 would be familiar enough about these, so you may want to add some pointers where he should read up to write about it.

I've adjusted the PR a bit to include GG's and my preferred changes. As you got some extra feedback, I'll hold back again from merging to give @nicolasperez19 a chance to respond.

ggerganov reviewed (approved):
It's a good first step - we'll expand after we merge it. Thanks

mofosyne commented:
@nicolasperez19 now merged in. Feel free to explore slaren suggestion in a separate PR
